pynamodb/attributes.py

```diff
@@ -257,3 +257,139 @@ def deserialize(self, value):
         """
         return parse(value, dayfirst=False).datetime
+
+
+class MapAttribute(Attribute):
+
+    attr_type = MAP
+    model = None
+
+    def __init__(self,
+                 hash_key=False,
+                 range_key=False,
+                 null=None,
+                 default=None,
+                 attr_name=None,
+                 model=None):
+        super(MapAttribute, self).__init__(hash_key=hash_key,
+                                           range_key=range_key,
+                                           null=null,
+                                           default=default,
+                                           attr_name=attr_name)
+        if model:
+            self.model = model
+
+    def validate_model(self, values):
+        if not self.model:
+            return True
+        return values.validate()
+
+    def serialize(self, values):
+        """
+        Encode the given list of numbers into a list of AttributeValue types.
+        """
+        if not self.validate_model(values):
+            raise Exception('model was not valid')
+        rval = dict()
+        invert_op = getattr(self, "__get_attributes", None)
+        if invert_op:
+            print 'inverting'
+            values = invert_op(values)
+        for k in values:
+            v = values[k]
+            if v is None:
+                rval[k] = {'NULL': bool(1)}
+            elif type(v) is dict:
+                rval[k] = {'M': MapAttribute().serialize(v)}
+            elif type(v) is list:
+                rval[k] = {'L': ListAttribute().serialize(v)}
+            elif type(v) is float or type(v) is int or type(v) is long:
+                rval[k] = {'N': NumberAttribute().serialize(v)}
+            elif type(v) is str or type(v) is unicode or type(v) is basestring:
+                rval[k] = {'S': UnicodeAttribute().serialize(v)}
+            elif type(v) is bool:
+                value_to_dump = bool(0)
+                if v:
+                    value_to_dump = bool(1)
+
+                rval[k] = {'BOOL': value_to_dump}
+            else:
+                raise Exception('Unknown value: ' + str(v))
+
+        return rval
+
+    def deserialize(self, values):
+        """
+        Decode numbers from list of AttributeValue types.
+        """
+        rval = dict()
+        for k in values:
+            v = values[k]
+            if 'L' in v:
+                rval[k] = ListAttribute().deserialize(v['L'])
+            elif 'N' in v:
+                rval[k] = NumberAttribute().deserialize(v['N'])
+            elif 'S' in v:
+                rval[k] = str(v['S'])
+            elif 'NULL' in v:
+                rval[k] = None
+            elif 'BOOL' in v:
+                rval[k] = bool(v['BOOL'])
+            elif 'M' in v:
+                rval[k] = MapAttribute().deserialize(v['M'])
+            else:
+                raise Exception('Unknown value: ' + str(v))
+        return rval
+
+
+class ListAttribute(Attribute):
```

Comment:
When a list is a list of maps, we can also let the list constructor take a model, and use that model to validate each member map.